flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentSchemaRegistryCoder.java

```diff
@@ -81,13 +100,29 @@ public Schema readSchema(InputStream in) throws IOException {
 
     @Override
     public void writeSchema(Schema schema, OutputStream out) throws IOException {
-        try {
-            int registeredId = schemaRegistryClient.register(subject, schema);
-            out.write(CONFLUENT_MAGIC_BYTE);
-            byte[] schemaIdBytes = ByteBuffer.allocate(4).putInt(registeredId).array();
-            out.write(schemaIdBytes);
-        } catch (RestClientException e) {
-            throw new IOException("Could not register schema in registry", e);
+        int registeredId;
+        if (registerSchema()) {
+            try {
+                registeredId = schemaRegistryClient.register(subject, schema);
+            } catch (RestClientException e) {
+                throw new IOException("Could not register schema in registry", e);
+            }
+        } else {
+            try {
+                registeredId = schemaRegistryClient.getId(subject, schema);
+            } catch (RestClientException e) {
+                throw new IOException("Could not retrieve schema in registry", e);
```

Comment:
Does it throw an IOException if the schema is not present?

Nit: Can we maybe throw a better exception e.g. FlinkException to make sure this is "expected" on not found schema?

Reply:
@fapaul this method implements an interface with a method that throws IOException, I think we should leave it throwing the IOException, so we can use the core interface. I see that the   `schemaRegistryClient.getId` throws an IOException, which then comes out of this method. We could wrap the IOException with a new IOException with more descriptive text, I have no strong views to do that. I would leave as coded by @MartijnVisser.

One improvement around the Exception processing would be to put the schema (though this could be big and might need to be truncating) or schema name and subject as inserts in the message.  WDYT?